### PR TITLE
[feat]:[CDS-81187]:[BE Changes-Fetch repositories for GAR]

### PR DESCRIPTION
**Pull request by Rakshitharness**
## Describe your changes
Created the new Api for GAR to fetch the repository.
Cases I have checked 

- V1 Call working fine 
- V2 call also working fine 

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55367)

### CODE REVIEW

File: 120-ng-manager/src/main/java/io/harness/ng/core/artifacts/resources/util/ArtifactResourceUtils.java [hunk 1]
```diff
@@ -1033,6 +1034,38 @@ public GARResponseDTO getBuildDetailsV2GAR(String gcpConnectorIdentifier, String
         resolvedPackage, version, versionRegex, orgIdentifier, projectIdentifier);
   }
 
+  public GARRepositoryDTO getRepositoriesV2GAR(String gcpConnectorIdentifier, String region, String project,
+      String accountId, String orgIdentifier, String pipelineIdentifier, String fqnPath, String serviceRef,
+      String runtimeInputYaml, String projectIdentifier, GitEntityFindInfoDTO gitEntityBasicInfo) {
+    if (StringUtils.isNotBlank(serviceRef)) {
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/core/artifacts/resources/util/ArtifactResourceUtils.java:1040]: Please refer other API implementations (getBuildDetailsV2GAR), we are considering remote services now.
Please check and update accordingly.

File: 120-ng-manager/src/main/java/io/harness/ng/core/artifacts/resources/gar/GARArtifactResource.java [hunk 2]
```diff
@@ -142,6 +143,42 @@ public class GARArtifactResource {
     return ResponseDTO.newResponse(buildDetails);
   }
 
+  @GET
+  @Path("getRepositories")
+  @ApiOperation(
+      value = "Gets Repositories google artifact registry", nickname = "getRepositoriesForGoogleArtifactRegistry")
+  public ResponseDTO<GARRepositoryDTO>
+  getRepositories(@QueryParam("connectorRef") String GCPConnectorIdentifier, @QueryParam("region") String region,
+      @QueryParam("project") String project, @NotNull @QueryParam(NGCommonEntityConstants.ACCOUNT_KEY) String accountId,
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/core/artifacts/resources/gar/GARArtifactResource.java:152]: please use constants in place of hardcoded value like "connectorRef"

File: 120-ng-manager/src/main/java/io/harness/ng/core/artifacts/resources/util/ArtifactResourceUtils.java [hunk 3]
```diff
@@ -1033,6 +1034,38 @@ public GARResponseDTO getBuildDetailsV2GAR(String gcpConnectorIdentifier, String
...
+      if (isBlank(gcpConnectorIdentifier)) {
+        gcpConnectorIdentifier = (String) googleArtifactRegistryConfig.getConnectorRef().fetchFinalValue();
+      }
+
+      if (isBlank(region)) {
+        region = (String) googleArtifactRegistryConfig.getRegion().fetchFinalValue();
+      }
+    }
+    gcpConnectorIdentifier =
+        getResolvedFieldValueWithYamlExpressionEvaluator(accountId, orgIdentifier, projectIdentifier,
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/core/artifacts/resources/util/ArtifactResourceUtils.java:1055]: In place of passing evaluator as null in method argument, please try re-using from previous response

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/artifact/resources/googleartifactregistry/dtos/GARRepositoryDTO.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import java.util.List;
+import lombok.AccessLevel;
+import lombok.Builder;
+import lombok.Value;
+import lombok.experimental.FieldDefaults;
+
+@CodePulse(module = ProductModule.CDS, unitCoverageRequired = true, components = {HarnessModuleComponent.CDS_ARTIFACTS})
+@Value
+@Builder
```
Comment (Abhishek-CDC) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/artifact/resources/googleartifactregistry/dtos/GARRepositoryDTO.java:23]: File names are interchanged. this should be dtoList and other one dto

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/googleartifactregistry/GARArtifactTaskHelper.java [hunk 5]
```diff
@@ -67,6 +67,13 @@ public ArtifactTaskResponse getArtifactCollectResponse(
               "Fetched Artifact details \n  type: Google Artifact Registry \n  Version: "
                   + garDelegateResponse.getVersion());
           break;
+        case GET_GAR_REPOSITORIES:
+          saveLogs(executionLogCallback, "Fetching artifact details");
+          artifactTaskResponse = getSuccessTaskResponse(garArtifactTaskHandler.getRepositories(attributes));
```
Comment (Abhishek-CDC) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/googleartifactregistry/GARArtifactTaskHelper.java:72]: we are not specifying what artifact details are being fetched

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/googleartifactregistry/GARArtifactTaskHandler.java [hunk 6]
```diff
@@ -97,6 +97,25 @@ public ArtifactTaskExecutionResponse getBuilds(GarDelegateRequest attributesRequ
...
+      garInternalConfig = getGarInternalConfig(attributesRequest);
+    } catch (IOException e) {
+      log.error("Could not get Bearer Token", e);
+      throw NestedExceptionUtils.hintWithExplanationException("Google Artifact Registry: Could not get Bearer Token",
+          "", new InvalidArtifactServerException(e.getMessage(), USER));
+    }
+    builds = garApiService.getRepository(garInternalConfig, attributesRequest.getRegion());
+    List<GarDelegateResponse> garArtifactDelegateResponseList =
+        builds.stream()
+            .sorted(new BuildDetailsInternalComparatorDescending())
```
Comment (Abhishek-CDC) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/googleartifactregistry/GARArtifactTaskHandler.java:114]: why we are adding this sort? as we don't have builds here

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java [hunk 7]
```diff
@@ -113,6 +114,18 @@ public List<BuildDetailsInternal> getBuilds(
...
   }
 
+  @Override
+  public List<BuildDetailsInternal> getRepository(GarInternalConfig garinternalConfig, String region) {
+    try {
+      GarRestClient garRestClient = getGarRestClient(garinternalConfig);
+      return getRepositories(garinternalConfig, garRestClient, region);
+    } catch (IOException e) {
+      throw NestedExceptionUtils.hintWithExplanationException("Could not fetch versions for the package",
+          "Please check if the package exists and if the permissions are scoped for the authenticated user",
```
Comment (Abhishek-CDC) [inline 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:124]: Please correct error message

File: 960-api-services/src/main/java/io/harness/artifacts/gar/GarRestClient.java [hunk 8]
```diff
@@ -37,4 +38,10 @@ Call<GarTags> getversioninfo(@Header("Authorization") String bearerAuthHeader,
       @Path(value = "project", encoded = true) String project, @Path(value = "region", encoded = true) String region,
       @Path(value = "repositories", encoded = true) String repositories, @Path(value = "package") String pkg,
       @Path(value = "tag", encoded = true) String version);
+
+  @GET("/v1/projects/{project}/locations/{region}/repositories")
+  Call<GarRepositoryResponse> getrepository(@Header("Authorization") String bearerAuthHeader,
+      @Path(value = "project", encoded = true) String project, @Path(value = "region", encoded = true) String region,
```
Comment (Abhishek-CDC) [inline 960-api-services/src/main/java/io/harness/artifacts/gar/GarRestClient.java:44]: please rename to getRepository

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java [hunk 9]
```diff
@@ -183,6 +196,42 @@ public BuildDetailsInternal verifyBuildNumber(GarInternalConfig garinternalConfi
...
+    List<BuildDetailsInternal> details = new ArrayList<>();
+    String project = garinternalConfig.getProject();
+    String nextPage = "";
+    do {
+      Response<GarRepositoryResponse> response =
+          garRestClient.getrepository(garinternalConfig.getBearerToken(), project, region, 1000, nextPage).execute();
+
+      if (response == null) {
+        throw NestedExceptionUtils.hintWithExplanationException("Response Is Null",
+            "Please Check Whether Artifact exists or not",
```
Comment (Abhishek-CDC) [inline 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:210]: Please improve error message

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java [hunk 10]
```diff
@@ -301,6 +350,40 @@ public List<BuildDetailsInternal> processPage(
...
 
+  public List<BuildDetailsInternal> processPage(
+      GarRepositoryResponse RepositoryPage, String versionRegex, GarInternalConfig garinternalConfig) {
+    if (RepositoryPage != null && EmptyPredicate.isNotEmpty(RepositoryPage.getRepositories())) {
+      int index = RepositoryPage.getRepositories().get(0).getName().lastIndexOf("/");
+      List<BuildDetailsInternal> buildDetails =
+          RepositoryPage.getRepositories()
+              .stream()
+              .map(repo -> {
+                String RepoName = repo.getName().substring(index + 1);
```
Comment (Abhishek-CDC) [inline 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:361]: variable name first letter small

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java [hunk 11]
```diff
@@ -301,6 +350,40 @@ public List<BuildDetailsInternal> processPage(
     }
   }
 
+  public List<BuildDetailsInternal> processPage(
+      GarRepositoryResponse RepositoryPage, String versionRegex, GarInternalConfig garinternalConfig) {
```
Comment (Abhishek-CDC) [inline 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:354]: versionRegex and garinternalConfig parameter is not required

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java [hunk 12]
```diff
@@ -301,6 +350,40 @@ public List<BuildDetailsInternal> processPage(
...
+                metadata.put("Format", RepoFormat);
+                metadata.put("createTime", createTime);
+                metadata.put("updateTime", UpdateTime);
+                return BuildDetailsInternal.builder().uiDisplayName(RepoName).metadata(metadata).build();
+              })
+              .filter(build
+                  -> StringUtils.isBlank(versionRegex) || new RegexFunctor().match(versionRegex, build.getNumber()))
+              .collect(toList());
+
+      return buildDetails.stream().sorted(new BuildDetailsInternalComparatorDescending()).collect(toList());
```
Comment (Abhishek-CDC) [inline 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:375]: this sort is on number and not repository

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java [hunk 13]
```diff
@@ -301,6 +350,40 @@ public List<BuildDetailsInternal> processPage(
...
+
+    } else {
+      if (RepositoryPage == null) {
+        log.warn("Google Artifact Registry Package version response was null.");
+      } else {
```
Comment (Abhishek-CDC) [inline 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:380]: incorrect logs

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java [hunk 14]
```diff
@@ -183,6 +196,42 @@ public BuildDetailsInternal verifyBuildNumber(GarInternalConfig garinternalConfi
...
+
+      GarRepositoryResponse page = response.body();
+      List<BuildDetailsInternal> pageDetails = processPage(page, null, garinternalConfig);
+      details.addAll(pageDetails);
+
+      if (page == null || StringUtils.isBlank(page.getNextPageToken())) {
+        break;
+      }
+
+      nextPage = StringUtils.isBlank(page.getNextPageToken()) ? null : page.getNextPageToken();
```
Comment (Abhishek-CDC) [inline 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:229]: Use StringUtils.defaultIfBlank() in place of ternary

Comment (Abhishek-CDC) [reply 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:229]: Why nextPage assigned null value initially if making null here

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java [hunk 15]
```diff
@@ -183,6 +196,42 @@ public BuildDetailsInternal verifyBuildNumber(GarInternalConfig garinternalConfi
...
+    } while (StringUtils.isNotBlank(nextPage));
+
+    return details.stream().sorted(new BuildDetailsInternalComparatorDescending()).collect(Collectors.toList());
```
Comment (Abhishek-CDC) [inline 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:232]: Sort can be here only and not in processPage.

Comment (Abhishek-CDC) [inline 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:232]: this sort is on number and not repository

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/googleartifactregistry/GARArtifactTaskHandler.java (hunk 6 above)
Comment (Abhishek-CDC) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/googleartifactregistry/GARArtifactTaskHandler.java:114]: Also sorting to be done on manager side. Please refer spec

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java (hunk 10 above)
Comment (Abhishek-CDC) [reply 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:361]: also extracting repoName from full path to be done on manager side. Please refer spec

**Comment by Rakshitharness** [comment]
trigger checkstylecodeformat

**Comment by Rakshitharness** [comment]
trigger messagecheck , trigger commonchecks

File: 120-ng-manager/src/main/java/io/harness/ng/core/artifacts/resources/gar/GARArtifactResource.java [hunk 16]
```diff
@@ -142,6 +143,45 @@ public class GARArtifactResource {
...
+    GARRepositoryDTOList modifiedBuildDetails =
+        artifactResourceUtils.getRepositoriesGAR(connectorRef, region, project, orgIdentifier, projectIdentifier);
+
+    return ResponseDTO.newResponse(modifiedBuildDetails);
+  }
+
+  @POST
+  @Path("v2/getRepositories")
+  @ApiOperation(
+      value = "Gets Repositories google artifact registry", nickname = "getRepositoriesForGoogleArtifactRegistry")
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/main/java/io/harness/ng/core/artifacts/resources/gar/GARArtifactResource.java:168]: i think having same nickname leads to swagger issue.
Please update and share swagger with @shivanand-harness to confirm

Comment (Abhishek-CDC) [reply 120-ng-manager/src/main/java/io/harness/ng/core/artifacts/resources/gar/GARArtifactResource.java:168]: we can update value also. please refer other APIs

File: 120-ng-manager/src/test/java/io/harness/ng/core/artifacts/resources/util/ArtifactResourceUtilsTest.java [hunk 17]
```diff
@@ -3413,6 +3416,150 @@ public void TestGetFilePathsS3() {
...
+        garResourceService.getRepositories(IDENTIFIER_REF, REGION, PROJECT, ORG_ID, PROJECT_ID);
+
+    int index = modifiedBuildDetails.getGarRepositoryDTOS().get(0).getRepository().lastIndexOf("/");
+    assertThat(modifiedBuildDetails).isNotNull();
+    assertThat(modifiedBuildDetails.getGarRepositoryDTOS()).hasSize(1);
+
+    GarRepositoryDTO modifiedRepo = modifiedBuildDetails.getGarRepositoryDTOS().get(0);
+    assertThat(modifiedRepo.getRepository().substring(index + 1)).isEqualTo("repo1");
+    assertThat(modifiedRepo.getFormat()).isEqualTo("format1");
+    assertThat(modifiedRepo.getCreateTime()).isEqualTo("2023-01-01T12:00:00Z");
```
Comment (Abhishek-CDC) [inline 120-ng-manager/src/test/java/io/harness/ng/core/artifacts/resources/util/ArtifactResourceUtilsTest.java:3559]: Please add UTs for config, queryParam, expression case similar to others

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/artifact/resources/googleartifactregistry/mappers/GARResourceMapper.java [hunk 18]
```diff
@@ -33,4 +35,20 @@ public static GARBuildDetailsDTO toGarBuildDetailsDTO(ArtifactBuildDetailsNG art
...
+                                                .map(response -> toGarRepositoryDTOList(response.getBuildDetails()))
+                                                .collect(Collectors.toList());
+    return GARRepositoryDTOList.builder().garRepositoryDTOS(detailsDTOList).build();
+  }
+
+  public static GarRepositoryDTO toGarRepositoryDTOList(ArtifactBuildDetailsNG artifactBuildDetailsNG) {
+    return GarRepositoryDTO.builder()
+        .repository(artifactBuildDetailsNG.getUiDisplayName())
+        .format(artifactBuildDetailsNG.getMetadata().get("Format"))
+        .createTime(artifactBuildDetailsNG.getMetadata().get("createTime"))
```
Comment (Abhishek-CDC) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/artifact/resources/googleartifactregistry/mappers/GARResourceMapper.java:50]: Please add these hardcoded values to ArtifactMetadataKeys interface and use

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java [hunk 19]
```diff
@@ -183,6 +197,42 @@ public BuildDetailsInternal verifyBuildNumber(GarInternalConfig garinternalConfi
...
+      Response<GarRepositoryResponse> response =
+          garRestClient.getRepository(garinternalConfig.getBearerToken(), project, region, 1000, nextPage).execute();
+
+      if (response == null) {
+        throw NestedExceptionUtils.hintWithExplanationException("Response Is Null",
+            "Please Check Whether repository exists for the given region:" + region,
+            new InvalidArtifactServerException(response.errorBody().toString(), USER));
+      }
+      if (!response.isSuccessful()) {
+        log.error("Request not successful. Reason: {}", response);
```
[the remark this answers is not in the document]
Comment (Abhishek-CDC) [reply 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:215]: we can make it warning in place of error

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java [hunk 20]
```diff
@@ -183,6 +197,42 @@ public BuildDetailsInternal verifyBuildNumber(GarInternalConfig garinternalConfi
...
+        if (!isSuccessful_Repository(response.code(), response.errorBody().toString())) {
+          throw NestedExceptionUtils.hintWithExplanationException("Unable to fetch Repository for the region:" + region,
+              "Please check region field", new InvalidArtifactServerException(response.message(), USER));
```
Comment (Abhishek-CDC) [inline 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:218]: we can omit field

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java [hunk 21]
```diff
@@ -301,6 +375,38 @@ public List<BuildDetailsInternal> processPage(
...
+          RepositoryPage.getRepositories()
+              .stream()
+              .map(repo -> {
+                String repoName = repo.getName();
+                String repoFormat = repo.getFormat();
+                String createTime = repo.getCreateTime();
+                String updateTime = repo.getUpdateTime();
+                Map<String, String> metadata = new HashMap();
+                metadata.put("Format", repoFormat);
+                metadata.put("createTime", createTime);
```
Comment (Abhishek-CDC) [inline 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:392]: Please add hardcoded values to ArtifactMetadataKeys interface and use

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/googleartifactregistry/GARArtifactTaskHandler.java [hunk 22]
```diff
@@ -97,6 +97,22 @@ public ArtifactTaskExecutionResponse getBuilds(GarDelegateRequest attributesRequ
...
+
+  public ArtifactTaskExecutionResponse getRepositories(GarDelegateRequest attributesRequest) {
+    List<BuildDetailsInternal> builds;
+    GarInternalConfig garInternalConfig;
+    try {
+      garInternalConfig = getGarInternalConfig(attributesRequest);
+    } catch (IOException e) {
+      log.error("Could not get Bearer Token", e);
+      throw NestedExceptionUtils.hintWithExplanationException("Google Artifact Registry: Could not get Bearer Token",
+          "", new InvalidArtifactServerException(e.getMessage(), USER));
```
Comment (Abhishek-CDC) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/googleartifactregistry/GARArtifactTaskHandler.java:109]: Please add UT

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java (hunk 21 above)
Comment (Rakshitharness) [reply 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:392]: Added values to ArtifactMetadatkeys

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java (hunk 13 above)
Comment (Rakshitharness) [reply 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:403]: Changes the logs

File: 120-ng-manager/src/main/java/io/harness/ng/core/artifacts/resources/util/ArtifactResourceUtils.java [hunk 23]
```diff
@@ -1033,6 +1036,75 @@ public GARResponseDTO getBuildDetailsV2GAR(String gcpConnectorIdentifier, String
...
+        garResourceService.getRepositories(connectorRef, region, project, orgIdentifier, projectIdentifier);
+
+    int index = buildDetails.getGarRepositoryDTOS().get(0).getRepository().lastIndexOf("/");
+
+    List<GarRepositoryDTO> modifiedBuilds = new ArrayList<>();
+
+    buildDetails.getGarRepositoryDTOS().forEach(repo -> {
+      String repoName = repo.getRepository().substring(index + 1);
+
+      GarRepositoryDTO modifiedRepo = GarRepositoryDTO.builder()
```
[the remark this answers is not in the document]
Comment (Rakshitharness) [reply 120-ng-manager/src/main/java/io/harness/ng/core/artifacts/resources/util/ArtifactResourceUtils.java:1095]: already added UT for this.

File: 120-ng-manager/src/test/java/io/harness/ng/core/artifacts/resources/util/ArtifactResourceUtilsTest.java (hunk 17 above)
Comment (Rakshitharness) [reply 120-ng-manager/src/test/java/io/harness/ng/core/artifacts/resources/util/ArtifactResourceUtilsTest.java:3559]: Added UT for these cases.

**Comment by Rakshitharness** [comment]
trigger uts

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (Rakshitharness) [reply 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java]: corrected the message

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java (hunk 20 above)
Comment (Rakshitharness) [reply 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java]: Done

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java [hunk 24]
```diff
@@ -183,6 +197,42 @@ public BuildDetailsInternal verifyBuildNumber(GarInternalConfig garinternalConfi
...
   }
 
+  private List<BuildDetailsInternal> getRepositories(GarInternalConfig garinternalConfig, GarRestClient garRestClient,
+      String region) throws WingsException, IOException {
+    List<BuildDetailsInternal> details = new ArrayList<>();
+    String project = garinternalConfig.getProject();
+    String nextPage = "";
+    do {
+      Response<GarRepositoryResponse> response =
+          garRestClient.getRepository(garinternalConfig.getBearerToken(), project, region, 1000, nextPage).execute();
```
[the remark this answers is not in the document]
Comment (Rakshitharness) [reply 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java]: Added the Constant For this

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/googleartifactregistry/GARArtifactTaskHandler.java (hunk 22 above)
Comment (Rakshitharness) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/googleartifactregistry/GARArtifactTaskHandler.java:109]: Added UT for This

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/artifact/resources/googleartifactregistry/mappers/GARResourceMapper.java (hunk 18 above)
Comment (Rakshitharness) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/artifact/resources/googleartifactregistry/mappers/GARResourceMapper.java]: Added hardcoded values to ArtifactMetadataKeys

File: 120-ng-manager/src/main/java/io/harness/ng/core/artifacts/resources/gar/GARArtifactResource.java (hunk 16 above)
Comment (Rakshitharness) [reply 120-ng-manager/src/main/java/io/harness/ng/core/artifacts/resources/gar/GARArtifactResource.java]: Updated the nickname and value

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java (hunk 14 above)
Comment (Rakshitharness) [reply 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java]: updated this to StringUtils.defaultIfBlank()

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java [hunk 25]
```diff
@@ -301,6 +375,38 @@ public List<BuildDetailsInternal> processPage(
...
+    String region = garinternalConfig.getRegion();
+    if (RepositoryPage != null && EmptyPredicate.isNotEmpty(RepositoryPage.getRepositories())) {
+      List<BuildDetailsInternal> buildDetails =
+          RepositoryPage.getRepositories()
+              .stream()
+              .map(repo -> {
+                String repoName = repo.getName();
+                String repoFormat = repo.getFormat();
+                String createTime = repo.getCreateTime();
+                String updateTime = repo.getUpdateTime();
```
[the remark this answers is not in the document]
Comment (Rakshitharness) [reply 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:392]: Already added

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java [hunk 26]
```diff
@@ -183,6 +198,44 @@ public BuildDetailsInternal verifyBuildNumber(GarInternalConfig garinternalConfi
...
+          garRestClient.getRepository(garinternalConfig.getBearerToken(), project, region, PAGESIZENEW, nextPage)
+              .execute();
+
+      if (response == null) {
+        throw NestedExceptionUtils.hintWithExplanationException("Response Is Null",
+            "Please Check Whether repository exists for the given region:" + region,
+            new InvalidArtifactServerException(response.errorBody().toString(), USER));
+      }
+      if (!response.isSuccessful()) {
+        log.warn("Request for Google Artifact Resource repository not successful for region:" + region + " Reason: {}",
```
Comment (Abhishek-CDC) [inline 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:217]: it is Registry, not resource

File: 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java [hunk 27]
```diff
@@ -301,6 +378,38 @@ public List<BuildDetailsInternal> processPage(
...
+              })
+              .collect(toList());
+
+      return buildDetails.stream().collect(toList());
+
+    } else {
+      if (RepositoryPage == null) {
+        log.warn("Google Artifact Registry repository response was null for the region:" + region);
+      } else {
+        log.warn("Google Artifact Registry repository response had an empty or missing tag list.");
```
Comment (Abhishek-CDC) [inline 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:407]: why we are saying missing tag list?

Comment (Abhishek-CDC) [reply 960-api-services/src/main/java/io/harness/artifacts/gar/service/GARApiServiceImpl.java:407]: Here we got empty list for repository

**Comment by Rakshitharness** [comment]
trigger uts

**Review by Abhishek-CDC — APPROVED** [review]

**Review by sarthakkasat — APPROVED** [review]

**Comment by Rakshitharness** [comment]
trigger review

**Pull request merged**